### ISSUE TITLE
ActionCable :if not respected?

### ISSUE THREAD

**Issue by donnfelker** [excid3/noticed#211]
I'm using JSP and I created a new notification with "rails g noticed:notification CommentNotification" and then added an `:if` to the `:action_cable` delivery method. 

Even though the `:if` is returning false, it seems to be sending the notification still. However, on my `:email` delivery method the if works properly. 

Any idea why this would happen? 

Notification: 
```ruby
# Notification when someone replies to a threaded comment
class CommentReplyNotification < ApplicationNotification
  # Database delivery is already added in ApplicationNotification
  deliver_by :action_cable, format: :to_websocket, channel: "NotificationChannel", if: :comment_from_someone_else?

  # Add required params
  #
  param :account
  param :comment

  # Define helper methods to make rendering easier.
  #
  # `message` and `url` are used for rendering in the navbar

  def message
    "#{comment.user.name} replied to your comment in the post: #{comment.commentable.title}"
  end

  def url
    # You can use any URL helpers here such as:
    # post_path(params[:post])
    polymorphic_path(comment.commentable)
  end

  # Include account_id to make sure notification only triggers if user is signed in to that account
  def to_websocket
    {
      account_id: record.account_id,
      html: ApplicationController.render(partial: "notifications/notification", locals: {notification: record})
    }
  end

  def comment_from_someone_else?
    false
    # even when hard coded this notification gets sent ... 🤔
    #recipient.id != comment.user.id
  end

  def comment
    params[:comment]
  end
end

```

**Comment by excid3** [comment]
Shoot, looks like this was accidentally removed in e244f66ec4b8640a118ec53cf04fa90812cf6f53

**Comment by excid3** [comment]
This got moved a while back so that it `if/unless` don't run immediately. The reason was you might want to delay a delivery method for 5 minutes (like email) but skip if it's read at that point.

The check now happens when the delivery method runs:
https://github.com/excid3/noticed/blob/master/lib/noticed/delivery_methods/base.rb#L48-L49

But for some reason this isn't working?

**Comment by donnfelker** [comment]
It's weird, it works for `:email` but for `:action_cable` the notification is still going through. 

This is not a huge thing for me, but I was going to give my users the ability to turn off "in app" notifications (notifcation bell in nav bar) but I've since decided against that as its non intrusive right now and I think having that on by default is useful for my app. So, this doesn't block me for now, but I could see it being an issue.

Do you know of a good way to test if action_cable received a call or not? I've got some ones for the `:email` delivery method by using ActionMailer::TestCase but I'm wondering if I can test the same with action cable (if/when I enable this feature).

**Comment by excid3** [comment]
Looking at the code, I don't see any reason why it would be any different. That's what's weird. 🤷‍♂️

There's probably something with the ActionCable::Channel that could be stubbed out for testing.

**Comment by omarluq** [comment]
@excid3 having a similar issue except that that the if condition is not being enforced for email or ios even when it returns false

**Comment by excid3** [comment]
Gonna need some way to reproduce the problem I think.

**Comment by omarluq** [comment]
@excid3 update! The if block is working, but even if an email or an iOS notification fails it still gets written into the delayed job queue, there is no way to capture a failed notification and no indication that a notification failed to start with. thats why I thought the if condition was not working! After adding loggers all over the place I was able to see if a notification went through or not and follow the lifecycle a little better. Would be nice if the response was available in the after_hook context, aside from that the gem been a blast to work with.

**Comment by excid3** [comment]
Correct. Because you can add delays, conditions are checked at time of delivery. That way you can skip sending an email notification if the user has already read it via ActionCable for example. 👍

**Comment by hsearcy** [comment]
I am running into this issue as well, but only in my test suite.
After putting a breakpoint [here](https://github.com/excid3/noticed/blob/master/lib/noticed/delivery_methods/base.rb#L47) inside `lib/noticed/delivery_methods/base`, I saw that the `@options` did not include the `:if` key. Upon further investigation, it was using the `deliver_by` options defined in my `ApplicationNotification` instead of the specific notification class.

Here are the relevant code snippets:

```
class ApplicationNotification < Noticed::Base
  # Delivery methods and helpers used by all notifications can be defined here.
  deliver_by :database, format: :to_database
  deliver_by :action_cable, channel: 'NotificationChannel', format: :to_action_cable
```  

```
class FileAddedNotification < ApplicationNotification
  deliver_by :action_cable, format: :to_action_cable, channel: 'NotificationChannel', delay: 5.seconds, if: :in_app_notifications?
```

And the spec calling it:
```
subject(:deliver_notification) { FileAddedNotification.with(account: account, file: file).deliver(users) }
```

I verified that this was indeed what was happening by adding the same `, if: :in_app_notifications?` and method into ApplicationNotification and adding a breakpoint inside the new method there. The breakpoint hit. Only during specs though... normal usage hit the intended code path.

To sum up the steps to debug after my spec failed:

1. Breakpoint in `FileAddedNotification.in_app_notifications?` -> does not get hit during specs.
2. Modify the `deliver_by :action_cable` in `ApplicationNotification` to use `if: some_check?` and define said method in `ApplicationNotification`
3. Breakpoint in `some_check?` gets hit on next spec run

As a workaround, doing this works:
```
class ApplicationNotification < Noticed::Base
  # Delivery methods and helpers used by all notifications can be defined here.
  deliver_by :database, format: :to_database
  deliver_by :action_cable, channel: 'NotificationChannel', format: :to_action_cable, if: :in_app_notifications?

  def in_app_notifications?
    true
  end
  ...
end
```

and then overriding `in_app_notifications` in `FileAddedNotification`
